src/eg/src/guardian_proof.rs

```diff
@@ -0,0 +1,276 @@
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct GuardianProof {
+    /// Guardian number, 1 <= i <= [`crate::varying_parameters::VaryingParameters::n`].
+    pub i: GuardianIndex,
+    /// Challenge vector
+    pub c: Vec<PossessionProofChallenge>,
+    /// Response vector
+    pub v: Vec<PossessionProofResponse>,
+}
+
+impl GuardianProof {
+    /// This function computes the challenge for the coefficient NIZK as specified in Equation (12)
+    /// The arguments are
+    /// - h_p - the parameter base hash
+    /// - i - the guardian index
+    /// - j - the coefficient index
+    /// - capital_k_i_j - the coefficient commitment
+    /// - h_i_j - the commit message
+    fn challenge(
+        h_p: HValue,
+        i: u32,
+        j: u32,
+        capital_k_i_j: &BigUint,
+        h_i_j: &BigUint,
+    ) -> PossessionProofChallenge {
+        // v = 0x10 | b(i,4) | b(j,4) | b(capital_k_i_j,512) | b(h_i_j,j,512)
+        let mut v = vec![0x10];
+        v.extend_from_slice(i.to_be_bytes().as_slice());
+        v.extend_from_slice(j.to_be_bytes().as_slice());
+        v.extend_from_slice(to_be_bytes_left_pad(capital_k_i_j, 512).as_slice());
+        v.extend_from_slice(to_be_bytes_left_pad(h_i_j, 512).as_slice());
+        let c = eg_h(&h_p, &v);
+        //The challenge is not reduced modulo q (cf. Section 5.4)
+        PossessionProofChallenge(BigUint::from_bytes_be(c.0.as_slice()))
+    }
+
+    /// This function computes coefficient NIZK proof as specified in Section 3.2.2
+    /// The arguments are
+    /// - csprng - secure randomness generator
+    /// - election_parameters - the election parameters
+    /// - h_p - the parameter base hash
+    /// - secret_key - the guardian `GuardianSecretKey`
+    pub fn new(
+        csprng: &mut Csprng,
+        election_parameters: &ElectionParameters,
+        h_p: HValue,
+        secret_key: &GuardianSecretKey,
+    ) -> Self {
+        let fixed_parameters = &election_parameters.fixed_parameters;
+        let varying_parameters = &election_parameters.varying_parameters;
+        let k = varying_parameters.k;
+
+        // Compute commit message
+        let u_vec = (0..k.get_one_based_u32())
+            .map(|_j| csprng.next_biguint_lt(fixed_parameters.q.borrow()))
+            .collect::<Vec<BigUint>>();
+        let h_vec = u_vec
+            .iter()
+            .map(|u_i| fixed_parameters.g.modpow(u_i, fixed_parameters.p.as_ref()))
+            .collect::<Vec<BigUint>>();
+        // Compute challenge vector
+        let c_vec = (0..k.get_one_based_u32())
+            .map(|j| {
+                Self::challenge(
+                    h_p,
+                    secret_key.i.get_one_based_u32(),
+                    j,
+                    &secret_key.coefficient_commitments.0[j as usize].0,
+                    &h_vec[j as usize],
+                )
+            })
+            .collect::<Vec<PossessionProofChallenge>>();
```

Comment:
```suggestion
        let c_vec = secret_key
            .coefficient_commitments
            .0
            .iter()
            .zip(&h_vec)
            .enumerate()
            .map(|(j, (coeff_comm, hj))| {
                Self::challenge(
                    h_p,
                    secret_key.i.get_one_based_u32(),
                    j as u32,
                    &coeff_comm.0,
                    &hj,
                )
            })
            .collect::<Vec<PossessionProofChallenge>>();
```

Take it or leave it :)

Reply:
Taken, see e.g. `guardian_secret_key.rs` :)